### DISCUSSION TITLE
Help with understanding how to setup Dynamics Module

### DISCUSSION THREAD

**Discussion by richmondodufisan** [idaholab/moose#26724] [Q&A Modules: Solid mechanics]
Hi,

I read through the Dynamics module page and think I understand it to some extent but I don't think I have it right as I am getting this error message:

```
*** ERROR ***
MooseVariableFE: Old time derivative of solution (`u_dot_old`) is not stored. Please set uDotOldRequested() to true in FEProblemBase before requesting `u_dot_old`.
```

I am currently trying to implement a simple cantilever beam subjected to a heaviside loading on the free end. First with no damping, and then eventually damping. After that I would like to add more features (plasticity, viscoelasticity, etc). I already implemented the simplest case (no damping) in MATLAB for a 1-D bar so I would like to confirm my results here. I'm using a 2D bar just so I can be able to visualize the traveling wave.

For the case with no damping, I attached my input file below:
[no_damp.zip](https://github.com/idaholab/moose/files/14146435/no_damp.zip)

```
[Mesh]
  type = FileMesh
  file = beam.msh
  use_displaced_mesh = false
[]

[GlobalParams]
  volumetric_locking_correction = false
  use_displaced_mesh = false
  displacements = 'disp_x disp_y'  
  order = FIRST
  family = LAGRANGE
[]

[Variables]
  [disp_x]
  []
  
  [disp_y]
  []
[]

[Kernels]
  [stress_x]
	type = DynamicStressDivergenceTensors
    variable = disp_x
    component = 0
	zeta = 0
  []
  [stress_y]
	type = DynamicStressDivergenceTensors
    variable = disp_y
    component = 1
	zeta = 0
  []
  [inertia_x]
    type = InertialForce
    variable = disp_x
	density = density
	eta = 0
  []
  [inertia_y]
    type = InertialForce
    variable = disp_y
	density = density
	eta = 0
  []
[]

[BCs]
  [right_y]
    type = DirichletBC
    variable = disp_y
    boundary = fixed_end
    value = 0.0
  []
  [right_x]
    type = DirichletBC
    variable = disp_x
    boundary = fixed_end
    value = 0.0
  []
  [left_x]
	type = FunctionNeumannBC
	variable = disp_x
	function = pressure
	boundary = loading_end
  []
[]

[Materials]
  [elast_tensor]
	type = ComputeIsotropicElasticityTensor
	youngs_modulus = 210e3
	poissons_ratio = 0.499
  []
  [strain]
    type = ComputeSmallStrain
  []
  [stress]
    type = ComputeLinearElasticStress
  []
  [density]
    type = GenericConstantMaterial
    prop_names = 'density'
    prop_values = '7.85e-6' # in kg/mm3
  []
[]

[Executioner]
  type = Transient
  solve_type = NEWTON
  nl_rel_tol = 1e-8
  nl_abs_tol = 1e-8
  dtmin = 1e-4
  timestep_tolerance = 1e-6
  start_time = -0.005
  end_time = 3750e-05
  dt = 2.445598573141631e-05
[]

[Functions]
  [pressure]
    type = ADParsedFunction
    expression = 'if(t<0,0.0,250)' # Heaviside function in MPa i.e N/mm2
  []
[]

[Outputs]
  exodus = true
  perf_graph = true
[]
```

-used DynamicStressDivergenceTensors for the stiffness term and used InertialForce for the Mass term. Also set both zeta and eta to zero explicitly. I am not sure what else is missing/why I am getting this error. I attached the input file and mesh below.

EDIT:
I "fixed" the issue by adding

```
  [TimeIntegrator]
    type = NewmarkBeta
    beta = 0.25
    gamma = 0.5
  []
```

To the executioner block. So apparently I needed to specify it directly. The documentation page says the transient problem is solved with HHT or Newmark which I'm guessing are the optimal ones to use. Would it be possible to use, say, Explicit Euler? That would involve trying to reconstruct the whole formulation, right?

**In addition to this, I have a few more questions:**
-How can I create an aux variable so I can see the stress distribution along the beam at the end of the simulation?

-Also, if I want to track the velocity and acceleration, I see that NewmarkAccelAux was used in other input files. How does that work? Does it automatically know how to calculate it through some relationship with DynamicStressDivergenceTensors?

-In 1-D, I don't need to worry about Poisson's ratio, but here I needed to set it to 0.499 (i.e nearly incompressible) to get the traveling wave solution, otherwise I got a really badly deformed mesh. I noticed while the simulation was running that it needed several iterations to solve. I have read about nearly incompressible solutions usually being problematic and less accurate (resolved through methods like the B-bar method). Is that a problem here? Because the next thing I want to do is to try to adjust the material properties so I can recreate the 1-D MATLAB solution.

Below is a screenshot of part of the solution:

![solution](https://github.com/idaholab/moose/assets/52802063/51ef6022-84fd-48c3-99a4-e839400d1a27)

the _disp value travels back and forth exactly like it did in 1D

**Comment by GiudGiud** [comment]
Hello

It s possible to use other time integrators but you need to modify them a little bit to request the uDotOld()

Guillaume

**Reply by richmondodufisan** [reply]
I see, thanks.

Also, to get the stress components as an aux variable (based on the loading I really only expect there to be xx stress), would it be:

```
  [stress_xx]
    order = CONSTANT
    family = MONOMIAL
  [xx]
```

in the AuxVariables block and

```
  [stress_xx]
    type = RankTwoAux
    variable = stress_xx
    rank_two_tensor = stress
    index_j = 0
    index_i = 0
    execute_on = timestep_end
  []
```

in the AuxKernels block? For stresses computed by ComputeLinearElasticStress, will the material property name ALWAYS be "stress"? What happens when I include plasticity? Will "stress" return the updated value including plasticity effects? 

***************************************************************************************************************************

As for the calculation of velocity and acceleration, which block actually does it? Is it the AuxKernels block i.e

```
  [./accel_x]
    type = NewmarkAccelAux
    variable = accel_x
    displacement = disp_x
    velocity = vel_x
    beta = 0.422
    execute_on = timestep_end
  [../]
```

or the Kernels block? i.e

```
  [./inertia_x]
    type = InertialForce
    variable = disp_x
    velocity = vel_x
    acceleration = accel_x
    beta = 0.422
    gamma = 0.8
    eta=0.1
    alpha = -0.3
  [../]
```

If the calculations are done in the AuxKernels block, why do we need to specify it in the Kernels block still? Also, how would I modify the whole script to use HHT Time Integration if I chose that instead?

***************************************************************************************************************************

Are there any checks implemented to prevent it from underpredicting the solution due to locking with poission's ratio = 0.499? Any fixes other than needing an incredibly fine mesh?

**Reply by GiudGiud** [reply]
Hello

For the first one, you can check the name of the material properties defined using Debug / show_material_props

The AuxKernel computes the quantity. The kernel consumes the acceleration variable.
I don't know about HHT. I imagine you ll want to use the same strategy as for Newmark, computing quantities on the side then having custom kernels that form the discretized equation

For the last question, I think there s a help page on volumetric locking in the module documentation

**Reply by richmondodufisan** [reply]
Thank you

For the middle question, why does the Kernel need to consume the acceleration & velocity variables if it can solve without it? What is it consuming it for? Those variables can be computed completely with just the displacements at every timestep (which is done in AuxKernel), can't they? So why still send to the Kernel?

**Reply by GiudGiud** [reply]
> why does the Kernel need to consume the acceleration & velocity variables if it can solve without it?

what makes you think it can solve without it? If you open the InertialForce.C file on your machine you can see what it does with those variables.

Being able to compute them everywhere does not mean we should recompute them everywhere. Sometimes it s faster to compute them once and re-use the result in several places
